java/arcs/core/storage/referencemode/MessageQueue.kt

```diff
@@ -45,27 +47,47 @@ class MessageQueue(
     @ExperimentalCoroutinesApi
     suspend fun enqueue(message: Message): Boolean {
         require(message !is Message.Enqueued) { "Cannot enqueue an already-enqueued message." }
-        size.incrementAndGet().let {
-            if (!sizeChannel.isClosedForSend) sizeChannel.send(it)
-        }
         val deferred = CompletableDeferred<Boolean>(coroutineContext[Job.Key])
-        mutex.withLock { queue += message.toEnqueued(deferred) }
+        mutex.withLock { queue += QueueEntry.MessageEntry(message.toEnqueued(deferred)) }
         CoroutineScope(coroutineContext).launch { drainQueue() }
-        return deferred.await().also {
-            size.decrementAndGet().let {
-                if (!sizeChannel.isClosedForSend) sizeChannel.send(it)
-            }
-        }
+        return deferred.await()
     }
 
     private suspend fun drainQueue() {
         val messagesToProcess = mutex.withLock {
-            mutableListOf<Message>().also {
+            mutableListOf<QueueEntry>().also {
                 it.addAll(queue)
                 queue.clear()
             }
         }
-        messagesToProcess.forEach { processMessage(it); Unit }
+
+        messagesToProcess.forEach {
+            when (it) {
+                is QueueEntry.MessageEntry -> processMessage(it.message)
+                is QueueEntry.Flush -> Unit
+            }
+        }
+
+        mutex.withLock {
+            if (queue.isEmpty()) {
+                onMessageQueueEmpty()
+            }
+        }
+    }
+
+    /**
+     * Posts a message on the queue, and returns a [Job] this will be completed when the queue
+     * arrives at this message. This allows you to wait for all *current* pending messages to
+     * be processed.
+     */
+    suspend fun flush(): Job {
+        val job = Job(coroutineContext[Job.Key])
+        mutex.withLock { queue += QueueEntry.Flush }
+        CoroutineScope(coroutineContext).launch {
+            drainQueue()
```

Comment:
ditto, the `drainQueue` can quickly arrive in `queue.isEmpty()` but there are still `processMessage(...)` in flight.

Reply:
Do you think it's reasonable to add a `activeQueueDrains` flag that's also set/cleared within the mutex, and used to decide whether or not empty condition is reached?

Or maybe a better solution is to serialize queue drains, so that they don't happen concurrently.

Reply:
I recommended refactoring the message queue stuff to channel/flow (the bug: b/159144117).
Multiple batches of processMessage(...) seems scary :-)
I'd prefer serializing queue drains to others like original atomicInt or flag.

Reply:
Great, I agree with this idea